Cancel button for product category block

mikejolley opened the pull request:
I wanted to have a quick go at #172 to see how far I could get. This is mainly for my own learning but hopefully I can get it good enough to merge :)

I will add inline comments for things I'm unsure of.

### Screenshots

![Edit Post ‹ local wordpress test — WordPress 2019-05-21 14-03-31](https://user-images.githubusercontent.com/90977/58098337-3d663300-7bd1-11e9-9e01-bbe31c7ccc97.png)

### How to test the changes in this Pull Request:

1. Edit a category block
2. Make some changes
3. Click cancel. Changes won't persist.
4. Edit again. Changes didn't persist. 
5. Make more changes. Click done. Changes did persist.

<!-- If you can, add the appropriate labels -->

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -37,8 +37,13 @@ class ProductByCategoryBlock extends Component {
 			products: [],
 			loaded: false,
+			changedAttributes: null,
+			isEditing: false,
```
I was looking at `ReusableBlockEdit` for inspiration and from past work I've seen this type of thing as state. I'm not sure why we wouldn't use state for editMode/isEditing - is there a reason I'm missing?

Aljullu replied:
That's a good question. @ryelle do you know if there are any pros/cons between using attributes or state for this?

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -48,6 +53,36 @@ class ProductByCategoryBlock extends Component {
+		} );
+	}
+
+	setAttributes( attributes ) {
```
I used `setAttributes` but I think there is a global/inherited setAttributes too.. I'm not sure where that is defined. Should I use another name? (this method updates the state, rather than the block attributes directly)

Aljullu replied:
Yes, `setAttributes` is one of the props this component receives, you can see props definition at the end of the file, under [`propTypes`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/583/files#diff-1f0006ba8a586d7d5f9939781d32257fR309).

I think it's usually better not to duplicate prop names as method names to avoid confusion (now we have `this.props.setAttributes` and `this.setAttributes`, which do different things but have the same name). For example, I would rename this method to `setChangedAttributes` or `updateChangedAttributes` or something like that.

ryelle replied:
I agree that this name `this.setAttributes` should be changed, either of the suggestions above work 👍

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -48,6 +53,36 @@ class ProductByCategoryBlock extends Component {
+		const { changedAttributes } = this.state;
+		const { setAttributes } = this.props;
+
+		setAttributes( changedAttributes );
```
This persists the changes afaik

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -85,23 +120,24 @@ class ProductByCategoryBlock extends Component {
 						operator={ catOperator }
 						onOperatorChange={ ( value = 'any' ) =>
-							setAttributes( { catOperator: value } )
+							this.setAttributes( { catOperator: value } )
```
This is calling my local state based setAttributes method. Am I calling this correctly? It works anyhow.

Aljullu replied:
It looks fine. As I said in another comment, I would change its name to avoid confusion with `setAttributes` prop and to make clear we are only updating the temporary attributes, but the calling looks good to me. :+1:

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -138,9 +174,20 @@ class ProductByCategoryBlock extends Component {
-		const { attributes, debouncedSpeak, setAttributes } = this.props;
+		const { attributes, debouncedSpeak } = this.props;
+		const { changedAttributes } = this.state;
+		const currentAttributes = { ...attributes, ...changedAttributes };
```
This is where I merge the unsaved attributes (from state) with the saved attributes (from this.props). I think I'm doing this correctly?

Aljullu replied:
I think so too. :slightly_smiling_face:

mikejolley commented on assets/js/blocks/product-category/block.js:
```diff
@@ -48,6 +53,36 @@ class ProductByCategoryBlock extends Component {
+	stopEditing() {
+		this.setState( {
+			isEditing: false,
+			changedAttributes: null,
```
This line clears any changes in the state that have not been persisted.

Aljullu commented on assets/js/blocks/product-category/block.js:
```diff
@@ -37,8 +37,13 @@ class ProductByCategoryBlock extends Component {
+		this.startEditing = this.startEditing.bind( this );
+		this.stopEditing = this.stopEditing.bind( this );
+		this.setAttributes = this.setAttributes.bind( this );
+		this.save = this.save.bind( this );
```
Correct. :+1: 

This is usually needed when passing the function as a callback (for example, when assigning the function to an `onClick` or similar attributes in a React component), because it loses the context and `this` would be `undefined`.

ryelle commented on assets/js/blocks/product-category/block.js:
```diff
@@ -161,33 +208,39 @@ class ProductByCategoryBlock extends Component {
 					</Button>
+					<Button
+						className="wc-block-products-category__cancel-button"
+						title={ __( 'Cancel', 'woo-gutenberg-products-block' ) }
```
Do you need the `title` prop? It's redundant for the accessibility case, since the button does contain text.

ryelle commented on assets/js/blocks/product-category/block.js:
```diff
@@ -220,14 +273,14 @@ class ProductByCategoryBlock extends Component {
 								title: __( 'Edit' ),
-								onClick: () => setAttributes( { editMode: ! editMode } ),
-								isActive: editMode,
+								onClick: () => this.startEditing(),
```
This should toggle the edit mode, not just "start editing" – probably when in edit mode, this should do the same behavior as cancelling out.

ryelle commented on assets/js/blocks/product-category/block.js:
```diff
@@ -161,33 +208,39 @@ class ProductByCategoryBlock extends Component {
+					<Button
+						className="wc-block-products-category__cancel-button"
+						title={ __( 'Cancel', 'woo-gutenberg-products-block' ) }
+						isLink
```
How about [`isTertiary` for text styled button instead?](https://github.com/wordpress/gutenberg/tree/master/packages/components/src/button#props) This link-styled button has an awkward focus state:

<img width="433" alt="Screen Shot 2019-05-22 at 3 05 28 PM" src="https://user-images.githubusercontent.com/541093/58201569-635c0800-7ca3-11e9-8f8b-bc0ed77b7815.png">

ryelle commented on assets/js/blocks/product-category/block.js:
```diff
@@ -48,6 +53,36 @@ class ProductByCategoryBlock extends Component {
+
+	setAttributes( attributes ) {
+		this.setState( ( prevState ) => {
+			if ( prevState.changedAttributes !== null ) {
```
I think this is preventing attributes from being set when the block is first added? At least, if I add the block, then try to select categories, they don't select.

Ah, I think it's because `changedAttributes` is `null` until you `startEditing` (and then it's `{}`), but `startEditing` is not called when the block is first added to the page.

mikejolley commented:
Thanks for the feedback. 🤦‍♂ I didn't even notice that sidebar panel.

ryelle commented:
I'm still triggering a state where the items can't be selected. If I clear all categories (next to 'X categories selected'), then click done, it resets into the editing state but the categories are un-selectable unless I click done/cancel and back into edit. ([screen recording](https://cloudup.com/cCGi9C2extG))

It's also a little odd to me that I can trigger the "editing" state by clearing categories from the sidebar, can select new categories there, but then need to click done in the main section to see the preview ([screen recording](https://cloudup.com/c62Oupc6Yvc))

I also think a designer should look through this, so I'm ccing @josemarques @jwold

mikejolley commented:
Thanks, I think I've tackled that now;

- Updating sidebar saves right away and overwrites and changes in edit mode
- Fixed the initial state so attributes can be changed

I do have to fix merge conflicts now however since the server side changes also changed the behaviour.

mikejolley commented:
The nothing found view was missing also. Putting it back.

jwold commented:
Just following up that I plan to take a look at this shortly. Thanks!

jwold commented:
Just reviewed the screenshots and looked at the original issue this is based off of. The main thing I was looking at was how the cancel button works, and adding that functionality looks good.

cc @mikejolley
